File: ros_tank_py/yaw_pid_node.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32, Bool
from geometry_msgs.msg import Twist, Pose
from tf_transformations import euler_from_quaternion
import time
import math as m
import numpy as np

class YawPIDController(Node):
    def __init__(self):
        super().__init__('yaw_pid_control')
        self.create_subscription(Float32, '/yaw_tgt', self.yaw_callback, 10)
        self.create_subscription(Pose, '/estimated_pose', self.pose_callback, 10)
        self.twist_pub = self.create_publisher(Twist, 'cmd_vel', 10)
        self.path_done_sub = self.create_subscription(Bool, '/path_done', self.path_done_callback, 10)

        # Tuned PID gains - start conservative
        self.Kp = 100 # Proportional gain
        self.Ki = 0 #0.05   # Integral gain
        self.Kd = 0 #0.2    # Derivative gain
        
        # PID limits
        self.max_integral = 1.0  # Anti-windup
        self.max_output = 3.0    # Max yaw rate output
        
        # System state
        self.target_yaw = 0.0
        self.current_yaw = 0.0
        self.integral = 0.0
        self.prev_error = 0.0
        self.prev_time = time.time()
        self.start_time = time.time()
        
        # Status flags
        self.yaw_tgt_rxd = False
        self.pose_rxd = False
        self.path_done = False
        
        # Low-pass filter for derivative term
        self.derivative_filter_alpha = 0.2  # Smoothing factor
        self.filtered_derivative = 0.0
        
        self.timer = self.create_timer(0.01, self.timer_callback)

    def normalize_angle(self, angle):
        """Normalize angle to [-π, π]"""
        return (angle + m.pi) % (2 * m.pi) - m.pi

    def path_done_callback(self, msg):
        self.path_done = msg.data

    def yaw_callback(self, msg):
        self.target_yaw = self.normalize_angle(msg.data)
        self.yaw_tgt_rxd = True

    def pose_callback(self, msg):
        self.pose_rxd = True
        _, _, self.current_yaw = euler_from_quaternion([
            msg.orientation.x,
            msg.orientation.y,
            msg.orientation.z,
            msg.orientation.w
        ])
        self.current_yaw = self.normalize_angle(self.current_yaw)

    def control_loop(self):
        if not self.yaw_tgt_rxd or not self.pose_rxd:
            return
        
        current_time = time.time()
        dt = current_time - self.prev_time
        self.prev_time = current_time

        if dt <= 0:
            return
            
        # Calculate error with angle wrapping
        error = self.normalize_angle(self.target_yaw - self.current_yaw)
        
        # Proportional term
        P = self.Kp * error
        
        # Integral term with anti-windup
        self.integral += error * dt
        self.integral = np.clip(self.integral, -self.max_integral, self.max_integral)
        I = self.Ki * self.integral
        
        # Derivative term with low-pass filtering
        raw_derivative = (error - self.prev_error) / dt
        self.filtered_derivative = (self.derivative_filter_alpha * raw_derivative + 
                                   (1 - self.derivative_filter_alpha) * self.filtered_derivative)
        D = self.Kd * self.filtered_derivative
        
        # Calculate output with saturation
        yaw_rate = P + I + D
        yaw_rate = np.clip(yaw_rate, -self.max_output, self.max_output)
        
        # Calculate forward velocity (optional)
        vx_magnitude = 0.25
        vx_tgt = vx_magnitude #* ((m.pi - abs(error)) / m.pi)
        
        # Publish command
        cmd_vel = Twist()
        cmd_vel.linear.x = vx_tgt
        cmd_vel.angular.z = yaw_rate

        if self.path_done:
           cmd_vel.linear.x = 0.0
           cmd_vel.angular.z = 0.0
           self.get_logger().info("Publishing zeros")
        self.twist_pub.publish(cmd_vel)
        # Update previous error
        self.prev_error = error
        
        # Logging
        #self.get_logger().info(
        #    f"Error: {error:.2f}, Yaw Rate: {yaw_rate:.2f}, "
        #    f"P: {P:.2f}, I: {I:.2f}, D: {D:.2f}"
        #)

    def timer_callback(self):
        self.control_loop()

def main(args=None):
    rclpy.init(args=args)
    node = YawPIDController()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__=='__main__':
    main()

File: ros_tank_cpp/yaw_pid_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float32.hpp>
#include <std_msgs/msg/bool.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <geometry_msgs/msg/pose.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2/LinearMath/Matrix3x3.h>
#include <chrono>
#include <cmath>
#include <algorithm>

class YawPIDController : public rclcpp::Node
{
public:
    YawPIDController() : Node("yaw_pid_control")
    {
        // Create subscriptions
        yaw_tgt_sub_ = this->create_subscription<std_msgs::msg::Float32>(
            "/yaw_tgt", 10, std::bind(&YawPIDController::yaw_callback, this, std::placeholders::_1));
        
        pose_sub_ = this->create_subscription<geometry_msgs::msg::Pose>(
            "/estimated_pose", 10, std::bind(&YawPIDController::pose_callback, this, std::placeholders::_1));
        
        path_done_sub_ = this->create_subscription<std_msgs::msg::Bool>(
            "/path_done", 10, std::bind(&YawPIDController::path_done_callback, this, std::placeholders::_1));

        // Create publisher
        twist_pub_ = this->create_publisher<geometry_msgs::msg::Twist>("cmd_vel", 10);

        // Tuned PID gains - start conservative
        Kp_ = 10.0;  // Proportional gain
        Ki_ = 0.0;    // Integral gain
        Kd_ = 0.0;    // Derivative gain

        // PID limits
        max_integral_ = 1.0;   // Anti-windup
        max_output_ = 3.0;     // Max yaw rate output

        // System state
        target_yaw_ = 0.0;
        current_yaw_ = 0.0;
        integral_ = 0.0;
        prev_error_ = 0.0;
        prev_time_ = std::chrono::steady_clock::now();
        start_time_ = std::chrono::steady_clock::now();

        // Status flags
        yaw_tgt_rxd_ = false;
        pose_rxd_ = false;
        path_done_ = false;

        // Low-pass filter for derivative term
        derivative_filter_alpha_ = 0.2;  // Smoothing factor
        filtered_derivative_ = 0.0;

        // Create timer (100 Hz)
        timer_ = this->create_wall_timer(
            std::chrono::milliseconds(10), std::bind(&YawPIDController::timer_callback, this));
    }

private:
    // Normalize angle to [-π, π]
    double normalize_angle(double angle)
    {
        return std::fmod(angle + M_PI, 2 * M_PI) - M_PI;
    }

    void path_done_callback(const std_msgs::msg::Bool::SharedPtr msg)
    {
        path_done_ = msg->data;
    }

    void yaw_callback(const std_msgs::msg::Float32::SharedPtr msg)
    {
        target_yaw_ = normalize_angle(msg->data);
        yaw_tgt_rxd_ = true;
    }

    void pose_callback(const geometry_msgs::msg::Pose::SharedPtr msg)
    {
        pose_rxd_ = true;
        
        // Convert quaternion to euler angles
        tf2::Quaternion quat(
            msg->orientation.x,
            msg->orientation.y,
            msg->orientation.z,
            msg->orientation.w
        );
        
        tf2::Matrix3x3 mat(quat);
        double roll, pitch, yaw;
        mat.getRPY(roll, pitch, yaw);
        
        current_yaw_ = normalize_angle(yaw);
    }

    void control_loop()
    {
        if (!yaw_tgt_rxd_ || !pose_rxd_)
            return;

        auto current_time = std::chrono::steady_clock::now();
        auto dt_chrono = current_time - prev_time_;
        double dt = std::chrono::duration<double>(dt_chrono).count();
        prev_time_ = current_time;

        if (dt <= 0)
            return;

        // Calculate error with angle wrapping
        double error = normalize_angle(target_yaw_ - current_yaw_);

        // Proportional term
        double P = Kp_ * error;

        // Integral term with anti-windup
        integral_ += error * dt;
        integral_ = std::clamp(integral_, -max_integral_, max_integral_);
        double I = Ki_ * integral_;

        // Derivative term with low-pass filtering
        double raw_derivative = (error - prev_error_) / dt;
        filtered_derivative_ = (derivative_filter_alpha_ * raw_derivative +
                               (1 - derivative_filter_alpha_) * filtered_derivative_);
        double D = Kd_ * filtered_derivative_;

        // Calculate output with saturation
        double yaw_rate = P + I + D;
        yaw_rate = std::clamp(yaw_rate, -max_output_, max_output_);

        // Calculate forward velocity (optional)
        double vx_magnitude = 0.25;
        double vx_tgt = vx_magnitude * ((M_PI - std::abs(error)) / M_PI);

        // Publish command
        auto cmd_vel = geometry_msgs::msg::Twist();
        cmd_vel.linear.x = vx_tgt;
        cmd_vel.angular.z = yaw_rate;

        if (path_done_)
        {
            cmd_vel.linear.x = 0.0;
            cmd_vel.angular.z = 0.0;
            RCLCPP_INFO(this->get_logger(), "Publishing zeros");
        }
        
        twist_pub_->publish(cmd_vel);
        
        // Update previous error
        prev_error_ = error;

        // Logging (commented out)
        // RCLCPP_INFO(this->get_logger(),
        //     "Error: %.2f, Yaw Rate: %.2f, P: %.2f, I: %.2f, D: %.2f",
        //     error, yaw_rate, P, I, D);
    }

    void timer_callback()
    {
        control_loop();
    }

    // Subscriptions
    rclcpp::Subscription<std_msgs::msg::Float32>::SharedPtr yaw_tgt_sub_;
    rclcpp::Subscription<geometry_msgs::msg::Pose>::SharedPtr pose_sub_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr path_done_sub_;
    
    // Publisher
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr twist_pub_;
    
    // Timer
    rclcpp::TimerBase::SharedPtr timer_;

    // PID gains
    double Kp_, Ki_, Kd_;
    
    // PID limits
    double max_integral_, max_output_;
    
    // System state
    double target_yaw_, current_yaw_;
    double integral_, prev_error_;
    std::chrono::steady_clock::time_point prev_time_, start_time_;
    
    // Status flags
    bool yaw_tgt_rxd_, pose_rxd_, path_done_;
    
    // Low-pass filter
    double derivative_filter_alpha_, filtered_derivative_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<YawPIDController>());
    rclcpp::shutdown();
    return 0;
}
